File: backend/app/schemas/report.py
"""
Report Pydantic schemas for request validation and response serialization.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.compliance.models import ComplianceReportStatus


class ReportBase(BaseModel):
    """Base schema containing common attributes for reports."""

    title: Optional[str] = Field(None, description="Optional title of the report")
    description: Optional[str] = Field(None, description="Optional description of the report")


class ReportCreate(ReportBase):
    """Schema for creating a new report."""

    pass


class ReportUpdate(BaseModel):
    """Schema for updating an existing report."""

    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = Field(None)


class ReportItemResponse(BaseModel):
    """Schema for report summary items returned in list/paginated endpoints."""

    id: uuid.UUID
    organization_id: uuid.UUID
    regulation_id: Optional[uuid.UUID] = None
    regulation_document_id: Optional[uuid.UUID] = None
    policy_document_id: Optional[uuid.UUID] = None
    overall_score: Optional[float] = None
    risk_level: Optional[str] = None
    total_matches: Optional[int] = None
    total_partial_matches: Optional[int] = None
    total_missing: Optional[int] = None
    report_status: ComplianceReportStatus
    created_at: datetime
    processing_time_seconds: Optional[float] = None
    processing_time_ms: Optional[float] = None
    is_deleted: bool = False
    open_count: Optional[int] = 0
    in_review_count: Optional[int] = 0
    remediation_count: Optional[int] = 0
    resolved_count: Optional[int] = 0

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class ReportPaginatedResponse(BaseModel):
    """Paginated response schema for GET /reports."""

    total: int = Field(..., description="Total number of matching reports")
    page: int = Field(..., description="Current page number (1-indexed)")
    page_size: int = Field(..., description="Number of items per page")
    items: List[ReportItemResponse] = Field(..., description="List of report summary items")


class ReportDetailResponse(BaseModel):
    """Detailed response schema for GET /reports/{report_id}."""

    id: uuid.UUID
    organization_id: uuid.UUID
    regulation_id: uuid.UUID
    regulation_document_id: Optional[uuid.UUID] = None
    policy_document_id: uuid.UUID
    overall_score: Optional[float] = None
    risk_level: Optional[str] = None
    summary: Optional[str] = None
    executive_summary: Optional[str] = None
    recommendations: Optional[Any] = None
    total_clauses: Optional[int] = None
    compliant_clauses: Optional[int] = None
    partial_clauses: Optional[int] = None
    non_compliant_clauses: Optional[int] = None
    total_matches: Optional[int] = None
    total_partial_matches: Optional[int] = None
    total_missing: Optional[int] = None
    processing_time_seconds: Optional[float] = None
    processing_time_ms: Optional[float] = None
    report_json: Optional[Any] = None
    report_status: ComplianceReportStatus
    is_deleted: bool = False
    created_at: datetime
    updated_at: datetime
    open_count: Optional[int] = 0
    in_review_count: Optional[int] = 0
    remediation_count: Optional[int] = 0
    resolved_count: Optional[int] = 0

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)



# Backwards compatibility alias
ReportResponse = ReportDetailResponse

File: client/src/types/report.ts
export type ComplianceReportStatus = 'PROCESSING' | 'COMPLETED' | 'FAILED' | 'PENDING';

export interface ReportItemResponse {
  id: string;
  organization_id: string;
  regulation_id?: string;
  regulation_document_id: string;
  policy_document_id: string;
  overall_score: number | null;
  risk_level?: string | null;
  total_matches?: number | null;
  total_partial_matches?: number | null;
  total_missing?: number | null;
  report_status: ComplianceReportStatus;
  created_at: string;
  processing_time_seconds: number | null;
  processing_time_ms?: number | null;
  is_deleted?: boolean;
}

export interface ReportPaginatedResponse {
  total: number;
  page: number;
  page_size: number;
  items: ReportItemResponse[];
}

export interface ReportDetailResponse {
  id: string;
  organization_id: string;
  regulation_id?: string;
  regulation_document_id: string;
  policy_document_id: string;
  overall_score: number | null;
  risk_level?: string | null;
  summary: string | null;
  executive_summary?: string | null;
  recommendations: Record<string, unknown> | Array<unknown> | null;
  total_clauses: number | null;
  compliant_clauses: number | null;
  partial_clauses: number | null;
  non_compliant_clauses: number | null;
  total_matches?: number | null;
  total_partial_matches?: number | null;
  total_missing?: number | null;
  processing_time_seconds: number | null;
  processing_time_ms?: number | null;
  report_json?: any;
  report_status: ComplianceReportStatus;
  is_deleted?: boolean;
  created_at: string;
  updated_at: string;
}

export interface ReportFilterParams {
  status?: string;
  orgSearch?: string;
  idSearch?: string;
  sortOrder?: 'newest' | 'oldest';
}
